analysis/deep2_r23_o32/plotting/plot_difference_curvefit.py

```diff
@@ -0,0 +1,284 @@
+import numpy as np
+from astropy.io import ascii as asc
+from scipy.optimize import curve_fit
+from os.path import join
+import matplotlib.pyplot as plt
+
+from Zcalbase_gal.analysis.deep2_r23_o32.log_commons \
+    import log_stdout, LogClass
+from Zcalbase_gal.analysis.deep2_r23_o32 import bian_coeff, ctype, \
+    secondorder_polynomial, thirdorder_polynomial, threevariable_fit, \
+    bian18_R23_OH, bian18_O32_OH, bian18_OH_O32, jiang_O32_OH_fit, jiang18_coeffs
+
+
+def plot_difference_threevariable(lR23, lO32, OH, lO32_all, pdf_file,
+                                  fitting_model, bin_start, bin_end,
+                                  new_coefficients=[], n_bins=4, data_err=[],
+                                  OH_err=[], OH_range=[7,11],
+                                  data_range=[-0.3, 0.3],
+                                  marker=[], label=[], IDs=[], log=None):
+    """
+    So this curve fitting function looks like it fits for three variables fits
+    a * x**3 + b*x**2 + c * x + d*lO32
+    This is also going to be used for the Jiang calibration
+
+
+    Plot differences between LACR23 vs observed R23 as
+    a function of metallicity
+    Plot differences between LACO32 vs observed O32 as
+    a function of metallicity
+    Used by main() function
+    """
+    # print('new coeff: ', new_coefficients, type(new_coefficients))
+    if log is None:
+        log = log_stdout()
+
+    # log.info("starting ...")
+
+    fig, ax = plt.subplots()
+
+    n_sample = len(lR23)
+
+    # Plotting
+    if len(marker) == 0:
+        marker = ['o'] * n_sample
+
+    diff0 = []
+    for nn in range(n_sample):
+        if fitting_model == 'jiang':
+            fitted_function = jiang_O32_OH_fit((OH[nn], lO32[nn]),
+                                               *jiang18_coeffs)
+            log.info(f"jiang_R23: {fitted_function}")
+        if fitting_model == 'Zcal':
+            if len(new_coefficients) != 0:
+                fitted_function = threevariable_fit((OH[nn], lO32[nn]),
+                                                    *new_coefficients)
+                # log.info(f"curve fit LAC_R23: {fitted_function}")
+
+        if nn == 0:
+            if IDs:
+                for jj in range(len(fitted_function)):
+                    id_diff = lR23[nn][jj] - fitted_function[jj]
+                    ax.annotate(IDs[nn][jj], (OH[nn][jj], id_diff),
+                                fontsize='6')
+
+        # Label in upper left the points
```

Comment:
This is labelling for each dataset (MACT, DEEP2, stacks), right?

Reply:
Yes it is the labeling for each dataset. The datasets are (Detections, Robust Limits, MACT, and DEEP2) so it currently only labels the detections.